apoc.hashing.fingerprintGraph fails if there are equals nodes

## Expected Behavior (Mandatory)

The function should return the hashsum.

## Actual Behavior (Mandatory)

The function throws a `Failed to invoke function `apoc.hashing.fingerprintGraph`: Caused by: java.lang.RuntimeException`

### Steps (Mandatory)

  1. `CREATE (:Person{name:'ABC'}), (:Person{name:'ABC'})`
  1. `RETURN apoc.hashing.fingerprintGraph()`

### Versions

  - OS: Mac BigSur
  - Neo4j: 4.3.3
  - Neo4j-Apoc: 4.3.0.3

--- a/core/src/main/java/apoc/hashing/Fingerprinting.java
+++ b/core/src/main/java/apoc/hashing/Fingerprinting.java
@@ -16,6 +16,7 @@ import org.neo4j.procedure.UserFunction;
 
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.AbstractMap;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Formatter;
@@ -24,6 +25,7 @@ import java.util.Map;
 import java.util.TreeMap;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
 public class Fingerprinting {
@@ -115,20 +117,15 @@ public class Fingerprinting {
             ));
 
             // step 2: build inverse map
-            Map<String, Long> nodeHashToId = idToNodeHash.entrySet().stream().collect(Collectors.toMap(
-                    Map.Entry::getValue,
-                    Map.Entry::getKey,
-                    (o, o2) -> {
-                        throw new RuntimeException();
-                    },
-                    () -> new TreeMap<>()
-            ));
+            final Map<String, List<Long>> nodeHashToId = idToNodeHash.entrySet()
+                    .stream()
+                    .collect(Collectors
+                            .groupingBy(Map.Entry::getValue, TreeMap::new,Collectors.mapping(Map.Entry::getKey, Collectors.toList())));
 
             // step 3: iterate nodes in order of their hash (we cannot rely on internal ids)
-            nodeHashToId.entrySet().stream().forEach(entry -> {
-                messageDigest.update(entry.getKey().getBytes());
-
-                Node node = tx.getNodeById(entry.getValue());
+            nodeHashToId.forEach((hash, ids) -> ids.forEach(id -> {
+                messageDigest.update(hash.getBytes());
+                Node node = tx.getNodeById(id);
                 List<EndNodeRelationshipHashTuple> endNodeRelationshipHashTuples = StreamSupport.stream(node.getRelationships(Direction.OUTGOING).spliterator(), false)
                         .map(relationship -> {
                             String endNodeHash = idToNodeHash.get(relationship.getEndNodeId());
@@ -140,8 +137,7 @@ public class Fingerprinting {
                     messageDigest.update(endNodeRelationshipHashTuple.getEndNodeHash().getBytes());
                     messageDigest.update(endNodeRelationshipHashTuple.getRelationshipHash().getBytes());
                 });
-
-            });
+            }));
 
         });
     }

--- a/core/src/test/java/apoc/hashing/FingerprintingTest.java
+++ b/core/src/test/java/apoc/hashing/FingerprintingTest.java
@@ -94,6 +94,17 @@ public class FingerprintingTest  {
     }
 
     @Test
+    public void fingerprintGraphShouldNotFailWithEqualsNodes() {
+        db.executeTransactionally("CREATE (:Person{name:'ABC'}), (:Person{name:'ABC'})");
+        String valueAfter = TestUtil.singleResultFirstColumn(db, "return apoc.hashing.fingerprintGraph()");
+        db.executeTransactionally("CREATE (:Person{name:'ABC'}), (:Person{name:'ABC'})");
+        String valueBefore = TestUtil.singleResultFirstColumn(db, "return apoc.hashing.fingerprintGraph()");
+        assertNotEquals(valueAfter, valueBefore);
+        String valueWithExclude = TestUtil.singleResultFirstColumn(db, "return apoc.hashing.fingerprintGraph(['name'])");
+        assertNotEquals(valueAfter, valueWithExclude);
+    }
+
+    @Test
     public void testExcludes() {
         compareGraph("CREATE (:Person{name:'ABC', created:timestamp()})", singletonList("created"), true);
     }

